Make getPath() virtual for PathOperation

nmoucht opened the pull request:
#### 493e2312e591137af161682ddbfab705c8132f7b
<pre>
Make getPath() virtual for PathOperation
<a href="https://bugs.webkit.org/show_bug.cgi?id=242882">https://bugs.webkit.org/show_bug.cgi?id=242882</a>

Reviewed by Simon Fraser.

Make getPath() virtual to get rid of need for getPathFromPathOperation.

* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::ReferencePathOperation::getPath const):
(WebCore::RayPathOperation::getPath const):
(WebCore::RayPathOperation::pathForReferenceRect const): Deleted.
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::setTransformInfo):
(WebCore::PathOperation::PathOperation):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyMotionPathTransform const):
(WebCore::getPathFromPathOperation): Deleted.

Canonical link: <a href="https://commits.webkit.org/252623@main">https://commits.webkit.org/252623@main</a>
</pre>

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1630,7 +1612,10 @@ void RenderStyle::applyMotionPathTransform(TransformationMatrix& transform, cons
     // Shift element to the point on path specified by offset-path and offset-distance.
-    auto path = getPathFromPathOperation(boundingBox, *offsetPath(), anchor, offsetRotate());
+    PathOperation* offset = offsetPath();
+    offset->setTransformInfo(boundingBox, anchor, offsetRotate());
```
It's not OK to mutate the PathOperation in the RenderStyle (because bits of RenderStyle can be shared between renderers).

nmoucht replied:
Ah I see. I made them function parameters with default values now.

smfr commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -50,6 +52,16 @@ const SVGElement* ReferencePathOperation::element() const
 
+const std::optional<Path> ReferencePathOperation::getPath(FloatRect reference, FloatPoint anchor, OffsetRotation rotation) const
+{
+    UNUSED_PARAM(anchor);
```
Instead of using `UNUSED_PARAM()` you can just remove the parameter name on line 55.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -118,14 +130,14 @@ double RayPathOperation::lengthForContainPath(const FloatRect& elementRect, doub
 }
 
-const Path RayPathOperation::pathForReferenceRect(const FloatRect& elementRect, const FloatPoint& anchor, const OffsetRotation rotation) const
+const std::optional<Path> RayPathOperation::getPath(FloatRect reference, FloatPoint anchor, OffsetRotation rotation) const
```
Call it `referenceRect`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -57,13 +57,12 @@ class PathOperation : public RefCounted<PathOperation> {
     OperationType type() const { return m_type; }
     bool isSameType(const PathOperation& o) const { return o.type() == m_type; }
-
+    virtual const std::optional<Path> getPath(FloatRect reference = FloatRect(), FloatPoint anchor = FloatPoint(), OffsetRotation rotation = OffsetRotation(false, 0)) const = 0;
```
I think we still prefer `const FloatRect&` for rect arguments (a FloatRect doesn't fit in a register).

You can also write this like:
`const FloatRect& reference = { }, FloatPoint anchor = { }, OffsetRotation rotation = { }` if you give `OffsetRotation` a default constructor.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -73,7 +72,7 @@ class ReferencePathOperation final : public PathOperation {
     const AtomString& fragment() const { return m_fragment; }
     const SVGElement* element() const;
-
+    const std::optional<Path> getPath(FloatRect reference = FloatRect(), FloatPoint anchor = FloatPoint(), OffsetRotation rotation = OffsetRotation(false, 0)) const override;
```
`final`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -103,6 +102,7 @@ class ShapePathOperation final : public PathOperation {
 
     void setReferenceBox(CSSBoxType referenceBox) { m_referenceBox = referenceBox; }
     CSSBoxType referenceBox() const { return m_referenceBox; }
+    const std::optional<Path> getPath(FloatRect reference = FloatRect(), FloatPoint anchor = FloatPoint(), OffsetRotation rotation = OffsetRotation(false, 0)) const override { UNUSED_PARAM(anchor); UNUSED_PARAM(rotation); return pathForReferenceRect(reference); }
```
`final`. You should not need the default values here, and you can just not name the parameters in this override.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -145,7 +145,7 @@ class BoxPathOperation final : public PathOperation {
     }
     
-    const Path getPath() const { return m_path; }
+    const std::optional<Path> getPath(FloatRect reference = FloatRect(), FloatPoint anchor = FloatPoint(), OffsetRotation rotation = OffsetRotation(false, 0)) const override { UNUSED_PARAM(anchor); UNUSED_PARAM(rotation); UNUSED_PARAM(reference); return m_path; }
```
Ditto

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -209,6 +208,7 @@ class RayPathOperation final : public PathOperation {
     {
         m_position = position;
     }
+    const std::optional<Path> getPath(FloatRect reference = FloatRect(), FloatPoint anchor = FloatPoint(), OffsetRotation rotation = OffsetRotation(false, 0)) const override;
```
`final`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -57,13 +57,12 @@ class PathOperation : public RefCounted<PathOperation> {
     OperationType type() const { return m_type; }
     bool isSameType(const PathOperation& o) const { return o.type() == m_type; }
-
+    virtual const std::optional<Path> getPath(const FloatRect& reference = { }, const FloatPoint anchor = { }, const OffsetRotation rotation = { }) const = 0;
```
We don't usually make by-value arguments `const`

nmoucht replied:
FloatRect needs to be const since it is a reference to a temporary. I made the rest non const.

smfr commented on Source/WebCore/rendering/style/OffsetRotation.h:
```diff
@@ -32,6 +32,11 @@ namespace WebCore {
 class OffsetRotation {
 public:
     OffsetRotation(bool hasAuto, float angle);
+    OffsetRotation()
```
This is fine, but you could also use initializers on the member variables and `OffsetRotation() = default`, or (preferably) you could give default values to the arguments to the existing constructor.

nmoucht replied:
Fixed.

nmoucht committed:
Make getPath() virtual for PathOperation
https://bugs.webkit.org/show_bug.cgi?id=242882

Reviewed by Simon Fraser.

Make getPath() virtual to get rid of need for getPathFromPathOperation.

* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::ReferencePathOperation::getPath const):
(WebCore::RayPathOperation::getPath const):
(WebCore::RayPathOperation::pathForReferenceRect const): Deleted.
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::setTransformInfo):
(WebCore::PathOperation::PathOperation):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyMotionPathTransform const):
(WebCore::getPathFromPathOperation): Deleted.

Canonical link: https://commits.webkit.org/252623@main
